README.md

```diff
@@ -16,25 +16,133 @@ specified bucket in GCS.
+combination determines the mode of record grouping the connector will
+use. Currently supported combinations of variables and the corresponding
+record grouping modes are:
+- `topic`, `partition`, `start_offset` - grouping by the topic and
+  partition;
+- `key` - grouping by the key.
+
+If the file name template is not specified, the default value is
 `{{topic}}-{{partition}}-{{start_offset}}` (+ `.gz` when compression is
 enabled).
 
-Even when `file.name.template` is explicitly set, the prefix is still
-added.
+### Record grouping
+
+Incoming records are being grouped until flushed.
+
+#### Grouping by the topic and partition
+
+In this mode, the connector groups records by the topic and partition.
+When a file is written, a offset of the first record in it is added to
+its name.
+
+For example, let's say the template is
+`t{{topic}}-p{{partition}}-o{{start_offset}}`. If the connector receives
+records like
+```
+topic:1 partition:0 offset:1
+topic:0 partition:0 offset:0
+topic:0 partition:0 offset:1
+topic:1 partition:0 offset:0
+flush
+```
+
+there will be two files `t0-p0-o0` and `t1-p0-o0` with two records in
+each.
+
+Each `flush` produces a new set of files. For example:
+
+```
+topic:0 partition:0 offset:0
+topic:0 partition:0 offset:1
+flush
+topic:0 partition:0 offset:2
+topic:0 partition:0 offset:3
+flush
+```
+
+In this case, there will be two files `t0-p0-o0` and `t1-p0-o2` with two
+records in each.
+
+#### Grouping by the key
+
+In this mode, the connector groups records by the Kafka key. It always
+puts one record in a file, the latest record that arrived before a flush
+for each key. Also, it overwrites files during if later new records with
+the same key arrives.
+
+The mode is good for maintaining the latest values per key as files on
+GCS.
+
+Let's say the template is `k{{key}}`. For example, when the following
+records arrive
+```
+key: 0 value: 0
+key: 1 value: 1
+key: 0 value: 2
+key: 1 value: 3
+flush
+```
+
+there will be two files `k0` (containing value `2`) and `k1` (containing
+value `3`).
+
+After a flush, previously written files might be overwritten:
+```
+key: 0 value: 0
+key: 1 value: 1
+key: 0 value: 2
+key: 1 value: 3
+flush
+key: 0 value: 4
+flush
+```
+
+In this case, there will be two files `k0` (containing value `4`) and
+`k1` (containing value `3`).
+
+##### The string representation of a key
+
+The connector in this mode uses the following algorithm to create the
+string representation of a key:
+
+1. If `key` is `null`, the string value is `"null"` (i.e., string
+   literal `null`).
+2. If `key` schema type is `STRING`, it's used directly.
+3. Otherwise, Java `.toString()` is applied.
+
+If keys of you records are strings, you may want to use
+`org.apache.kafka.connect.storage.StringConverter` as `key.converter`.
+
+##### Warning: Single key in different partitions
+
+This mode primarily targets the situation where each key appears in one
+partition.
```

Comment:
This is a bit hard to understand (does this refer to some new mode or what). Maybe something like this instead?

> The `group by key` mode primarily targets scenarios where each key appears in one partition only. If the same key appears in multiple partitions the result may be unexpected.